File: Data Structures & Algorithms/combinations/submission-1.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []
        path = [] # [0] * k 
        def dfs(i):
            d = k - len(path)
            if i < d:
                return
            if len(path) == k:
                res.append(path[:])
                return
            for j in range(i, 0, -1):
                path.append(j)
                dfs(j -1)
                path.pop()
        dfs(n)
        return res

File: Data Structures & Algorithms/palindrome-partitioning/submission-1.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> res = new ArrayList<>();
        List<String> path = new ArrayList<>();
        dfs(0, s, path, res);
        return res;
    }

    private void dfs(int i, String s, List<String> path, List<List<String>> res) {
        if (i == s.length()) {
            res.add(new ArrayList<>(path));
            return;
        }

        for (int j = i; j < s.length(); j++) {
            if (isPalindorme(s, i, j)) {
                path.add(s.substring(i, j + 1));
                dfs(j + 1, s, path, res);
                path.removeLast();
            }
        }
    }

    private boolean isPalindorme(String s, int left, int right) {
        while (left < right) {
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
}
